File: skill-test/projects/sample-cluster-java/src/test/java/cluster/java/ClusterActorTest.java
package cluster.java;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.cluster.Cluster;
import akka.cluster.MemberStatus;
import akka.testkit.javadsl.TestKit;
import com.typesafe.config.ConfigFactory;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import scala.concurrent.duration.FiniteDuration;
import java.util.concurrent.TimeUnit;
import java.util.stream.StreamSupport;

class ClusterActorTest {
    private static ActorSystem system;

    @BeforeAll
    static void setup() {
        system = ActorSystem.create("ClusterTestSystem", ConfigFactory.load());
        // Single-node cluster: join self
        Cluster cluster = Cluster.get(system);
        cluster.join(cluster.selfAddress());
        awaitClusterUp(system, 1, Duration.ofSeconds(10));
    }

    @AfterAll
    static void teardown() {
        TestKit.shutdownActorSystem(system,
            FiniteDuration.apply(10, TimeUnit.SECONDS), true);
    }

    private static void awaitClusterUp(ActorSystem targetSystem, int expectedMembers, Duration timeout) {
        new TestKit(targetSystem) {{
            awaitAssert(timeout, Duration.ofMillis(200), () -> {
                long upCount = StreamSupport.stream(
                        Cluster.get(targetSystem).state().getMembers().spliterator(), false)
                    .filter(m -> m.status().equals(MemberStatus.up()))
                    .count();

                if (upCount < expectedMembers) {
                    throw new AssertionError("Expected at least " + expectedMembers + " Up members but got " + upCount);
                }
                return null;
            });
        }};
    }

    @Test
    void clusterListenerReceivesMemberUpEvent() {
        new TestKit(system) {{
            ActorRef listener = system.actorOf(
                ClusterListenerActor.props(getRef()), "listener-test");

            // Should receive MemberUp notification for self node
            expectMsgEquals(Duration.ofSeconds(5), "member-up");
            expectNoMessage(Duration.ofMillis(500));
        }};
    }

    @Test
    void counterSingletonActorCountsCorrectly() {
        new TestKit(system) {{
            ActorRef counter = system.actorOf(CounterSingletonActor.props(), "counter-test");

            counter.tell(CounterSingletonActor.Increment.INSTANCE, getRef());
            counter.tell(CounterSingletonActor.Increment.INSTANCE, getRef());
            counter.tell(CounterSingletonActor.Increment.INSTANCE, getRef());
            counter.tell(new CounterSingletonActor.GetCount(getRef()), getRef());

            CounterSingletonActor.CountValue result =
                expectMsgClass(Duration.ofSeconds(3), CounterSingletonActor.CountValue.class);
            assert result.getValue() == 3 : "Expected count 3, got " + result.getValue();
        }};
    }

    @Test
    void pubSubDeliversMessageToSubscriber() {
        new TestKit(system) {{
            // Create subscriber with reportTo = testkit probe
            ActorRef subscriber = system.actorOf(
                PubSubSubscriberActor.props("test-topic", getRef()), "sub-test");

            // Wait for subscription acknowledgment
            expectMsgEquals(Duration.ofSeconds(5), "subscribed");

            // Create publisher and publish
            ActorRef publisher = system.actorOf(PubSubPublisherActor.props(), "pub-test");
            publisher.tell("hello-cluster", ActorRef.noSender());

            // Subscriber should forward message to reportTo
            expectMsgEquals(Duration.ofSeconds(5), "hello-cluster");
            expectNoMessage(Duration.ofMillis(500));
        }};
    }
}

File: skill-test/projects/sample-cluster-kotlin/src/test/kotlin/cluster/kotlin/ClusterActorTest.kt
package cluster.kotlin

import com.typesafe.config.ConfigFactory
import org.apache.pekko.actor.testkit.typed.javadsl.ActorTestKit
import org.apache.pekko.actor.typed.ActorRef
import org.apache.pekko.cluster.MemberStatus
import org.apache.pekko.cluster.typed.Cluster
import org.apache.pekko.cluster.typed.Join
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicInteger

class ClusterActorTest {
    companion object {
        private lateinit var testKit: ActorTestKit

        @JvmStatic
        @BeforeAll
        fun setup() {
            val config = ConfigFactory.parseString("""
                pekko {
                  actor.provider = "cluster"
                  remote.artery {
                    canonical.hostname = "127.0.0.1"
                    canonical.port = 0
                  }
                  cluster {
                    seed-nodes = []
                    min-nr-of-members = 1
                    downing-provider-class = "org.apache.pekko.cluster.sbr.SplitBrainResolverProvider"
                  }
                  loglevel = "INFO"
                }
            """.trimIndent()).withFallback(ConfigFactory.load())

            testKit = ActorTestKit.create(config)
            // Single-node cluster: join self
            val cluster = Cluster.get(testKit.system())
            cluster.manager().tell(Join.create(cluster.selfMember().address()))
            awaitClusterUp(testKit, 1, Duration.ofSeconds(10))
        }

        @JvmStatic
        @AfterAll
        fun teardown() {
            testKit.shutdownTestKit()
        }

        private fun awaitClusterUp(testKit: ActorTestKit, expectedMembers: Int, timeout: Duration) {
            val observer = testKit.createTestProbe<String>()
            observer.awaitAssert(timeout, Duration.ofMillis(200)) {
                val upCount = Cluster.get(testKit.system()).state().members.count {
                    it.status() == MemberStatus.up()
                }
                require(upCount >= expectedMembers) {
                    "Expected at least $expectedMembers Up members but got $upCount"
                }
                "cluster-ready"
            }
        }
    }

    @Test
    fun `counter singleton actor counts correctly`() {
        val counter = testKit.spawn(CounterSingletonActor.create(), "counter-test")
        val probe = testKit.createTestProbe<CounterCommand>()

        counter.tell(Increment)
        counter.tell(Increment)
        counter.tell(Increment)
        counter.tell(GetCount(probe.ref()))

        val result = probe.expectMessageClass(CountValue::class.java)
        assert(result.value == 3) { "Expected 3, got ${result.value}" }
    }

    @Test
    fun `pubsub delivers message to subscriber`() {
        val pubSubManager = testKit.spawn(PubSubManagerActor.create(), "pubsub-test")
        val subscriberProbe = testKit.createTestProbe<String>()

        // Subscribe
        pubSubManager.tell(SubscribeToTopic("test-topic", subscriberProbe.ref()))

        subscriberProbe.awaitAssert(Duration.ofSeconds(5), Duration.ofMillis(200)) {
            pubSubManager.tell(PublishMessage("test-topic", "hello-cluster"))
            subscriberProbe.expectMessage(Duration.ofMillis(700), "hello-cluster")
            "delivered"
        }
    }

    @Test
    fun `kafka singleton should execute runner only once`() {
        val calls = AtomicInteger(0)
        val fakeRunner = object : KafkaStreamRunner {
            override fun runOnce(
                bootstrapServers: String,
                topic: String,
                groupIdPrefix: String,
                payload: String,
                timeout: Duration
            ) = CompletableFuture.completedFuture(payload).also { calls.incrementAndGet() }
        }

        val singleton = testKit.spawn(
            KafkaStreamSingletonActor.create(
                "localhost:9092",
                "cluster-kotlin-events",
                "test-group",
                Duration.ofSeconds(3),
                fakeRunner
            ),
            "kafka-singleton-test"
        )

        singleton.tell(StartKafkaStream)
        singleton.tell(StartKafkaStream)
        singleton.tell(StartKafkaStream)

        val observer = testKit.createTestProbe<String>()
        observer.awaitAssert(Duration.ofSeconds(3), Duration.ofMillis(100)) {
            kotlin.test.assertEquals(1, calls.get())
            "done"
        }
    }

    @Test
    fun `kafka singleton should stop gracefully on stop message`() {
        val singleton = testKit.spawn(
            KafkaStreamSingletonActor.create(
                "localhost:9092",
                "cluster-kotlin-events",
                "test-group",
                Duration.ofSeconds(3)
            ),
            "kafka-singleton-stop-test"
        )

        val watcher = testKit.createTestProbe<Any>()
        singleton.tell(StopKafkaStream)
        watcher.expectTerminated(singleton, Duration.ofSeconds(3))
    }
}
